miniscope_io/models/devupdate.py

```diff
@@ -0,0 +1,123 @@
+"""
+Models for device update commands.
+"""
+
+from enum import Enum
+
+import serial.tools.list_ports
+from pydantic import BaseModel, ConfigDict, field_validator, model_validator
+
+
+class DeviceCommand(Enum):
+    """Commands for device."""
+
+    REBOOT = 200
+
+
+class UpdateCommandDefinitions:
+    """
+    Definitions of Bit masks and headers for remote update commands.
+    """
+
+    # Header to indicate target/value.
+    # It probably won't be used in other places so defined here.
+    id_header = 0b00000000
+    target_header = 0b11000000
+    LSB_header = 0b01000000
+    MSB_header = 0b10000000
+    LSB_value_mask = 0b000000111111  # value below 12-bit
+    MSB_value_mask = 0b111111000000  # value below 12-bit
+    reset_byte = 0b11111111
+
+
+class UpdateTarget(Enum):
+    """
+    Targets to update. Needs to be under 6-bit.
+    """
+
+    LED = 0
```

Comment:
using enums as mappings to meaningful values like this is usually more trouble than an enum with opaque values and a separate value mapping (e.g. needing to get the value in the validator), but don't want to be too nitpicky. fine with me just giving my experience :)

Reply:
I'm not sure if I get your intention here. Would you mind making this change so I can see through your example?

Reply:
it can be somewhat fragile to treat the value of an enum as meaningful, at least in python - i think a good rule of thumb is to treat them like you would RDBMS `id` columns or persistent identifiers, opaque unique values that identify an object but don't contain information so that the information can freely change without the identifier needing to change.

thinking about this situation recently: https://github.com/linkml/linkml-runtime/pull/310#issuecomment-2011023357

where the value of an enum was meaningful, and then over time it became the case accidentally that multiple enum keys got the same value (which doesn't work and invisibly caused problems) and it wasn't really possible to reuse those enum keys and give them further information as became necessary.

so the idea of an enum is good here, we do want to have some representation of the possible kinds of commands that can be given, and i think it would be even better if this was an enum for all possible kinds of commands across all miniscopes. that might not have been what you had in mind here, but we will need something like that eventually and this seems as good a candidate as any.

one might imagine wanting to do something like this:

```python
WirelessMiniscope.configure(UpdateTarget.GAIN, 10)
WirelessMiniscopeV2.configure(UpdateTarget.GAIN, 10)
AnotherMiniscope.configure(UpdateTarget.GAIN, 10)
EphysDevice.configure(UpdateTarget.GAIN, 10)
I2CDevice.configure(UpdateTarget.GAIN, 10)
# ... and so on
```

if the value of the enum is meaningful beyond being a unique identifier though, then...
- we wouldn't be able to e.g. modify the firmware and use a different value, 
- or we would have to be careful to ensure that the value was the same across all the firmwares of all devices, 
- or we would have to have mappings between the enum value and the specific value that some different firmware uses to set GAIN
- we would have trouble adapting it to different modalities like i2c where the value might be some hex register address

we would also sacrifice some ergonomics that might be nice to have, e.g. we might want to allow someone to just pass a string like "`miniscope.set('gain', 10)`" instead of needing to import the enum (while also constraining the possible values that can be accepted, so not just having the signature of `set` be `set(key: str, ...)` but instead be something like `set(key: SomeKeyEnum, ...)`). 

So we might do something like this:

```python
class UpdateTarget(str, Enum):
    # behaves similarly to inheriting from int and enum, 
    # part of stdlib `enum` as of 3.11
    LED = "led"
    GAIN = "gain"
    # ...
    
class SomeMiniscope(Device):
    command_map: ClassVar[dict[UpdateTarget, int]] = {
        UpdateTarget.LED: 0,
        UpdateTarget.GAIN: 1,
        # ...
    } 
    
    def set(self, key: UpdateTarget, value: Any):
        # say this is some hardware abstraction class like OkDev...
        self._inner_hardware.set(self.command_map[key], value)

```

where the value of the enum is still useful as a string, but carries no further information aside from being one of a set of unique values. 

again, doesn't have to be in this PR, and also this might not be what this class is for, but just explaining what i was thinking in previous comment

Reply:
Ah, I get your point. Thanks for the detailed guide.
It makes sense that using `enum` this way is vulnerable to accidents as we extend this in multiple ways, but it also still feels pretty arbitrary regarding how we should implement this to me, so I will skip it for now.

Reply:
that is fair. it is maybe premature to think about uniformity of interface vs. heterogeneity of implementation when we don't yet have a uniform interface.